lib/rust/ensogl/core/src/system/gpu/shader/compiler.rs

```diff
@@ -0,0 +1,396 @@
+use crate::system::gpu::shader::Fragment;
+use crate::system::gpu::shader::Shader;
+use crate::system::gpu::shader::Vertex;
+use crate::system::web;
+
+use web_sys::WebGl2RenderingContext;
+
+
+
+// =================
+// === Constants ===
+// =================
+
+/// We do not want the framerate to drop below this value when compiling shaders. Whenever we
+/// discover that it drops below this threshold, no more shaders will be scheduled for compilation
+/// or linking in this frame. It does not mean, however, that the framerate will not be lower. The
+/// last scheduled shader can take a lot of time to be rendered, causing the FPS to drop
+/// significantly.
+const MIN_FPS: f32 = 60.0;
+const MAX_FRAME_TIME_MS: f32 = 1000.0 / MIN_FPS;
+
+
+
+// ===========
+// === Job ===
+// ===========
+
+/// Compiler job. After the job is created it can be either transformed to another job, or, in case
+/// that was the final job, the [`handler`] callback will be called. See the documentation of the
+/// [`Compiler`] to learn more.
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Job<T> {
+    #[derivative(Debug = "ignore")]
+    on_ready: Box<dyn FnOnce(shader::Program)>,
+    input:    T,
+    handler:  WeakJobHandler,
+}
+
+/// A handler to a job. After the handler is dropped, the job is invalidated and will no longer be
+/// scheduled for evaluation.
+#[derive(Debug, Clone, CloneRef)]
+pub struct JobHandler {
+    rc: Rc<()>,
+}
+
+/// A weak version of [`JobHandler`].
+#[derive(Debug, Clone, CloneRef)]
+pub struct WeakJobHandler {
+    weak: Weak<()>,
+}
+
+impl JobHandler {
+    /// Constructor.
+    fn new() -> Self {
+        Self { rc: default() }
+    }
+
+    /// Get weak reference to this handler.
+    pub fn downgrade(&self) -> WeakJobHandler {
+        let weak = Rc::downgrade(&self.rc);
+        WeakJobHandler { weak }
+    }
+}
+
+impl WeakJobHandler {
+    /// heck whether the handler was dropped.
+    pub fn exists(&self) -> bool {
+        self.weak.upgrade().is_some()
+    }
+}
+
+
+
+// ================
+// === Compiler ===
+// ================
+
+/// Compiler job queues. See the documentation of [`Compiler`] to learn more.
+#[derive(Debug, Default)]
+pub struct Jobs {
+    compile:              Vec<Job<shader::Code>>,
+    link:                 Vec<Job<shader::CompiledCode>>,
+    khr_completion_check: Vec<Job<shader::Program>>,
+    link_check:           Vec<Job<shader::Program>>,
+}
+
+/// Compiles and links GL shader programs, asynchronously. The compiler works in the following way:
+/// 1. A new shader code is submitted with the [`submit`] method. A new job is created in the
+///    [`Jobs::compile`] queue and the job handler is returned to the user.
+///
+/// 2. The following pseudo-algorithm is performed:
+/// ```text
+/// on_ever_frame if job queues are not empty {
+///     Promote ready jobs from [`Jobs::khr_completion_check`] to [`Jobs::link_check`].
+///     Loop while the current frame time is < MAX_FRAME_TIME_MS {
```

Comment:
Repeat question of the comment from above: so on a weak system, we might never compile any shaders?

Reply:
Good catch. The docs did not explain it properly. fixed the docs.